include/ignition/gazebo/comms/ICommsModel.hh

```diff
@@ -0,0 +1,202 @@
+  ///                      messages. The default value is "/broker/msgs"
+  ///    <bind_service>: Service name used to bind an address.
+  ///                    The default value is "/broker/bind"
+  ///    <unbind_service>: Service name used to unbind from an address.
+  ///                      The default value is "/broker/unbind"
+  ///
+  /// Here's an example:
+  /// <physics name="1ms" type="ignored">
+  ///   <max_step_size>2</max_step_size>
+  ///   <real_time_factor>1.0</real_time_factor>
+  /// </physics>
+  /// <plugin
+  ///   filename="ignition-gazebo-perfect-comms-system"
+  ///   name="ignition::gazebo::systems::PerfectComms">
+  ///   <step_size>1</step_size>
+  /// </plugin>
+  class ICommsModel
+      : public System,
+        public ISystemConfigure,
+        public ISystemPreUpdate
+  {
+    /// \brief Destructor.
+    public: virtual ~ICommsModel() = default;
+
+    // Documentation inherited.
+    public: void Configure(const Entity &_entity,
+                           const std::shared_ptr<const sdf::Element> &_sdf,
+                           EntityComponentManager &_ecm,
+                           EventManager &_eventMgr) override
+    {
+      sdf::ElementPtr sdfClone = _sdf->Clone();
+
+      // Parse the optional <step_size>.
+      if (sdfClone->HasElement("step_size"))
+      {
+        this->timeStep = std::chrono::duration<int64_t, std::nano>(
+          static_cast<int64_t>(sdfClone->Get<double>("step_size") * 1e9));
+      }
+
+      this->Load(_entity, sdfClone, _ecm, _eventMgr);
+      this->broker.Load(sdfClone);
+      this->broker.Start();
+    }
+
+    // Documentation inherited.
+    public: void PreUpdate(
+                const ignition::gazebo::UpdateInfo &_info,
+                ignition::gazebo::EntityComponentManager &_ecm) override
+    {
+      IGN_PROFILE("ICommsModel::PreUpdate");
+
+      if (_info.paused)
+        return;
+
+      this->currentTime =
+        std::chrono::steady_clock::time_point(_info.simTime);
+
+      if (!this->timeStep.has_value())
+      {
+        // If no step_size is defined simply execute one step of the comms model
+        this->StepImpl(_info, _ecm);
+      }
+      else
+      {
+        // Otherwise step at the specified time step until we converge on the
+        // final timestep. If the timestep is larger than the dt, then dt will
+        // be used.
+        auto endTime = this->currentTime + _info.dt;
+
+        while (this->currentTime < endTime)
+        {
+          ignition::gazebo::UpdateInfo info(_info);
+          info.dt = std::min(this->timeStep.value(), _info.dt);
+          info.simTime = this->currentTime.time_since_epoch();
+          this->StepImpl(_info, _ecm);
+          this->currentTime += info.dt;
+        }
+      }
+    }
+
+    /// \brief This method is called when there is a timestep in the simulator
+    /// override this to update your data structures as needed.
```

Comment:
Comment `override this to update your data structures as needed.` is at both `Step` and `StepImpl`. Which one should the user override, then?

Reply:
Clarified in [1140b56](https://github.com/ignitionrobotics/ign-gazebo/pull/1416/commits/1140b565c175cbdae114a39158f4639a373b04db).